homeassistant/components/tplink/light.py

```diff
@@ -117,436 +72,108 @@ def brightness_from_percentage(percent):
-            monthly_statistics = self.smartbulb.get_emeter_monthly()
-            try:
-                self._emeter_params[ATTR_DAILY_ENERGY_KWH] = round(
-                    float(daily_statistics[int(time.strftime("%d"))]), 3
-                )
-                self._emeter_params[ATTR_MONTHLY_ENERGY_KWH] = round(
-                    float(monthly_statistics[int(time.strftime("%m"))]), 3
-                )
-            except KeyError:
-                # device returned no daily/monthly history
-                pass
-
-    async def _async_set_light_state_retry(
-        self, old_light_state: LightState, new_light_state: LightState
-    ) -> None:
-        """Set the light state with retry."""
-        # Tell the device to set the states.
-        if not _light_state_diff(old_light_state, new_light_state):
-            # Nothing to do, avoid the executor
-            return
-
-        self._is_setting_light_state = True
-        try:
-            light_state_params = await self.hass.async_add_executor_job(
-                self._set_light_state, old_light_state, new_light_state
-            )
-            self._is_available = True
-            self._is_setting_light_state = False
-            if LIGHT_STATE_ERROR_MSG in light_state_params:
-                raise HomeAssistantError(light_state_params[LIGHT_STATE_ERROR_MSG])
-            # Some devices do not report the new state in their responses, so we skip
-            # set here and wait for the next poll to update the values. See #47600
-            if LIGHT_STATE_ON_OFF in light_state_params:
-                self._light_state = self._light_state_from_params(light_state_params)
-            return
-        except (SmartDeviceException, OSError):
-            pass
-
-        try:
-            _LOGGER.debug("Retrying setting light state")
-            light_state_params = await self.hass.async_add_executor_job(
-                self._set_light_state, old_light_state, new_light_state
-            )
-            self._is_available = True
-            if LIGHT_STATE_ERROR_MSG in light_state_params:
-                raise HomeAssistantError(light_state_params[LIGHT_STATE_ERROR_MSG])
-            self._light_state = self._light_state_from_params(light_state_params)
-        except (SmartDeviceException, OSError) as ex:
-            self._is_available = False
-            _LOGGER.warning("Could not set data for %s: %s", self.smartbulb.host, ex)
-
-        self._is_setting_light_state = False
-
-    def _set_light_state(
-        self, old_light_state: LightState, new_light_state: LightState
-    ) -> None:
-        """Set the light state."""
-        diff = _light_state_diff(old_light_state, new_light_state)
+        return SUPPORT_TRANSITION
 
-        if not diff:
-            return
+    @property
+    def supported_color_modes(self) -> set[str] | None:
+        """Return list of available color modes."""
+        modes = set()
```

Comment:
modes should not be an empty set, if nothing else is supported it should contain COLOR_MODE_ONOFF. Please review the developer docs again, maybe cross reference with other integrations. Or there are no TPLink binary lights?

Reply:
Ah, the documentation is not that clear on this instance - it discusses COLOR_MODE_ONOFF only on legacy cases where `color_mode` is not supported. However, I modified it to add that mode if none other have been matched (which should not happen though).